transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -548,27 +548,23 @@ private static Encoding toEncoding(@Nullable String value) {
   @Override
   public void recordLogEventDropped(
       long eventsDroppedCount, LogEventDropped.Reason reason, String logSource) {
-    if (!isLogEventDroppedExist(reason, logSource)) {
-      ContentValues metrics = new ContentValues();
-      metrics.put("log_source", logSource);
-      metrics.put("reason", reason.getNumber());
-      metrics.put("events_dropped_count", eventsDroppedCount);
-      inTransaction(
-          db -> {
-            db.insert("log_event_dropped", null, metrics);
-            return null;
-          });
-    } else {
-      String query =
-          "UPDATE log_event_dropped SET events_dropped_count = events_dropped_count + "
-              + eventsDroppedCount
-              + " WHERE log_source = ? AND reason = ?";
-      inTransaction(
-          db -> {
-            db.execSQL(query, new String[] {logSource, Integer.toString(reason.getNumber())});
-            return null;
-          });
-    }
+    String sql =
+        "INSERT INTO log_event_dropped (log_source, reason, events_dropped_count) "
+            + "VALUES(?, ?, ?) "
+            + "ON CONFLICT(log_source, reason) "
+            + "DO UPDATE SET events_dropped_count = events_dropped_count + ?";
```

Comment:
It's really cool to see upsert, but I am not sure we can afford to use it. i.e. https://www.sqlite.org/lang_UPSERT.html mentions that it was added in version 3.24.0. While we support devices with level all the way down to 16+ https://stackoverflow.com/a/4377116

So we would have to do two queries like you used to do, but place them in one transaction.

Reply:
didn't realize API 26 and API 27 [link](https://developer.android.com/reference/android/database/sqlite/package-summary) are using a bit older version of SQLite, but i will change it.